openforcefield/tests/test_parameters.py

```diff
@@ -292,6 +292,60 @@ class MyParameterType(_ParameterAttributeHandler, MixIn):
 
 class TestParameterHandler:
 
+    from simtk import unit
+    length = 1*unit.angstrom
+    k = 10*unit.kilocalorie_per_mole/unit.angstrom**2
+
+    def test_add_parameter(self):
+        """Test the behavior of add_parameter"""
+        bh = BondHandler(skip_version_check=True)
+        param1 = {'smirks': '[*:1]-[*:2]', 'length': self.length, 'k': self.k, 'id': 'b1'}
+        param2 = {'smirks': '[*:1]=[*:2]', 'length': self.length, 'k': self.k, 'id': 'b2'}
+        param3 = {'smirks': '[*:1]#[*:2]', 'length': self.length, 'k': self.k, 'id': 'b3'}
+
+        bh.add_parameter(param1)
+        bh.add_parameter(param2)
+        bh.add_parameter(param3)
+
+        assert [param.id for param in bh._parameters] == ['b1', 'b2', 'b3']
+
+        param_duplicate_smirks = {'smirks': param2['smirks'], 'length': 2*self.length, 'k': 2*self.k}
+
+        with pytest.raises(ValueError):
+            bh.add_parameter(param_duplicate_smirks)
+
+        dict_to_add_by_smirks = {'smirks': '[#1:1]-[#6:2]', 'length': self.length, 'k': self.k, 'id': 'd1'}
+        dict_to_add_by_index = {'smirks': '[#1:1]-[#8:2]', 'length': self.length, 'k': self.k, 'id': 'd2'}
+
+        param_to_add_by_smirks = BondHandler.BondType(
+            **{'smirks': '[#6:1]-[#6:2]', 'length': self.length, 'k': self.k, 'id': 'p1'}
+        )
+        param_to_add_by_index = BondHandler.BondType(
+            **{'smirks': '[#6:1]=[#8:2]', 'length': self.length, 'k': self.k, 'id': 'p2'}
+        )
+
+        with pytest.raises(ValueError):
```

Comment:
I found it easy to lose track of why we expected an error here. Could you add comments for each failure mode in this test to explain why the failure happens?